ext/liquid_c/variable_lookup.c

```diff
@@ -2,63 +2,34 @@
 #include "context.h"
 
 static ID id_has_key, id_aref, id_fetch;
-static ID id_ivar_lookups, id_ivar_name, id_ivar_command_flags;
 
-VALUE variable_lookup_evaluate(VALUE self, VALUE context)
+VALUE variable_lookup_key(VALUE context, VALUE object, VALUE key, bool is_command)
 {
-    long command_flags = -1;
-    VALUE name = rb_ivar_get(self, id_ivar_name);
-    name = context_evaluate(context, name);
-
-    VALUE object = context_find_variable(context, name, Qtrue);
-
-    VALUE lookups = rb_ivar_get(self, id_ivar_lookups);
-    Check_Type(lookups, T_ARRAY);
-
-    for (long i = 0; i < RARRAY_LEN(lookups); i++) {
-        VALUE key = context_evaluate(context, RARRAY_AREF(lookups, i));
-        VALUE next_object;
-
-        if (rb_respond_to(object, id_aref) && (
-            (rb_respond_to(object, id_has_key) && rb_funcall(object, id_has_key, 1, key)) ||
-            (rb_obj_is_kind_of(key, rb_cInteger) && rb_respond_to(object, id_fetch))
-        )) {
-            next_object = rb_funcall(object, id_aref, 1, key);
-            next_object = materialize_proc(context, object, key, next_object);
-            object = value_to_liquid_and_set_context(next_object, context);
-            continue;
-        }
+    if (rb_respond_to(object, id_aref) && (
+        (rb_respond_to(object, id_has_key) && rb_funcall(object, id_has_key, 1, key)) ||
+        (rb_obj_is_kind_of(key, rb_cInteger) && rb_respond_to(object, id_fetch))
+    )) {
+        VALUE next_object = rb_funcall(object, id_aref, 1, key);
+        next_object = materialize_proc(context, object, key, next_object);
+        return value_to_liquid_and_set_context(next_object, context);
+    }
 
-        if (command_flags == -1) {
-            command_flags = NUM2LONG(rb_ivar_get(self, id_ivar_command_flags));
-        }
-        if (command_flags & (1 << i)) {
-            Check_Type(key, T_STRING);
-            ID intern_key = rb_intern(RSTRING_PTR(key));
-            if (rb_respond_to(object, intern_key)) {
-                next_object = rb_funcall(object, rb_intern(RSTRING_PTR(key)), 0);
-                object = value_to_liquid_and_set_context(next_object, context);
-                continue;
-            }
+    if (is_command) {
+        Check_Type(key, T_STRING);
+        ID intern_key = rb_intern(RSTRING_PTR(key));
```

Comment:
Yeah it seems like we could avoid the conversion through RString here as mentioned in another comment. Definitely not a blocker, tiny optimization.

Reply:
We could optimize commands for this case, but we would still need to pass a string as the key to `rb_funcall(object, id_aref, 1, key)` if the object is a hash.